conn_browser.go

```diff
@@ -0,0 +1,292 @@
+// +build js,wasm
+
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errConnectionClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	errorHandler   *js.Func
+	mut            sync.Mutex
+	currDataMut    sync.RWMutex
+	currData       bytes.Buffer
+	closeOnce      sync.Once
+	closeSignal    chan struct{}
+	dataSignal     chan struct{}
+	localAddr      net.Addr
+	remoteAddr     net.Addr
+	firstErr       error
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:       raw,
+		closeSignal: make(chan struct{}),
+		dataSignal:  make(chan struct{}, 1),
+		localAddr:   NewAddr("0.0.0.0:0"),
+		remoteAddr:  getRemoteAddr(raw),
+	}
+	// Force the JavaScript WebSockets API to use the ArrayBuffer type for
+	// incoming messages instead of the Blob type. This is better for us because
+	// ArrayBuffer can be converted to []byte synchronously but Blob cannot.
+	conn.Set("binaryType", "arraybuffer")
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
+		return c.readAfterErr(b)
+	}
+
+	for {
+		c.currDataMut.RLock()
+		n, err := c.currData.Read(b)
+		c.currDataMut.RUnlock()
+		if err != nil && err != io.EOF {
+			// Return any unexpected errors immediately.
+			return n, err
+		} else if n == 0 || err == io.EOF {
+			// There is no data ready to be read. Wait for more data or for the
+			// connection to be closed.
+			select {
+			case <-c.dataSignal:
+				continue
+			case <-c.closeSignal:
+				return c.readAfterErr(b)
+			}
+		} else {
+			return n, err
+		}
+	}
+}
+
+// readAfterError reads from c.currData. If there is no more data left it
+// returns c.firstErr if non-nil and otherwise returns io.EOF.
+func (c *Conn) readAfterErr(b []byte) (int, error) {
+	c.currDataMut.RLock()
+	n, err := c.currData.Read(b)
+	c.currDataMut.RUnlock()
+	if n == 0 {
+		if c.firstErr != nil {
+			return 0, c.firstErr
+		} else {
+			return 0, io.EOF
+		}
+	}
+	return n, err
+}
+
+// checkOpen returns an error if the connection is not open. Otherwise, it
+// returns nil.
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errConnectionClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	if err := c.checkOpen(); err != nil {
```

Comment:
Will something go wrong if we drop this check and just try writing? This will race with any `Close` calls anyways.

Reply:
@Stebalien according to [this GitHub issue](https://github.com/websockets/ws/issues/1515) browsers do not throw an exception if you try to call `send` on a closed connection. You can verify for yourself by visiting https://www.websocket.org/echo.html and running the following in the JavaScript console:

```
ws = new WebSocket("wss://echo.websocket.org");
ws.onopen = () => {
  ws.close();
  ws.send("test");
  console.log("sent");
};
```

Chrome prints a message to the console but it doesn't throw an exception. Firefox doesn't give any indication that anything is wrong.

I think we have to leave this check in place just so that callers can know if they are trying to send to a closed connection. It _should_ be okay if it's racy because the connection will be detected as closed pretty quickly in any case. What do you think?

Reply:
Then this _should_ be fine.